repo structure update plus web api project added

HoroshoNaVolge opened the pull request.

HoroshoNaVolge committed:
repo structure update

HoroshoNaVolge committed:
AddAspNetCoreWebApi

for use with Dadata Api.

navferty commented on src/AspNetCore.WebApi/appsettings.json:
```diff
@@ -0,0 +1,14 @@
+
+  "DaData": {
+    "DaDataApiBaseUrl": "http://suggestions.dadata.ru/suggestions/api/4_1/rs/findById/party/",
+    "DaDataApiToken": "bbc8e57adbc7bebaf5265017740c30fce634cdc0"
```
Один из вариантов, который может позволить не раскрывать ключи и иную чувствительную информацию в системе контроля версий - это использование `appsettings.Development.json`. Этот файл, с одной стороны, при локальном запуске в дебаг режиме, переопределяет значения, которые были в `appsettings.json`, а с другой стороны - может быть добавлен в гитигнор, и исключен из системы контроля версий.
Так как файл уже был закоммичен, можно либо переписать историю коммитов (я бы пока рекомендовал так глубоко не копать), либо сделать так:
* удалить файл appsettings.Development.json из репозитория (ну или просто перенести его в другую папку вне репы, чтобы не пришлось создавать заново), и закомиттить это удаление
* вернуть его обратно, но добавить в гитигнор (вручную, либо в GUI студии ПКМ->ignore this item в панели гит)

navferty commented on src/AspNetCore.WebApi/Program.cs:
```diff
@@ -0,0 +1,52 @@
+    {
+        public static void Main(string[] args)
+        {
+            Other.AutofacLearn.Run();
```
Я бы предложил выделить в другой проект эксперименты с autofaq, а в текущем приложении сосредоточиться на использовании доступного "из коробки" контейнера microsoft.extensions.dependencyinjection

navferty commented on src/AspNetCore.WebApi/Program.cs:
```diff
@@ -0,0 +1,52 @@
+        {
+            Other.AutofacLearn.Run();
+
+            var builder = WebApplication.CreateBuilder(args);
```
Если не ошибаюсь, этот метод должен по умолчанию добавлять некоторую конфигурацию, в том числе джсон файлы appsettings.json и appsettings.{EnvironmentName}.json

Можно проверить, поковыряв исходный код [тыц](https://github.com/dotnet/aspnetcore/blob/main/src/DefaultBuilder/src/WebApplication.cs), [тыц](https://github.com/dotnet/aspnetcore/blob/main/src/DefaultBuilder/src/WebApplicationBuilder.cs#L34), или запустить и посмотреть, подцепит ли он конфигу из файлов.

navferty replied:
Видимо я не раскрыл до конца свою мысль. Я имел в виду, что так как этот билдер уже по умолчанию добавляет в том числе эти файлы (наряду с некоторыми другими источниками конфигурации, например, переменными среды), можно убрать явное добавление этих двух файлов (тем более что appsettings.Development.json и не должен добавляться, если окружение - не Development).

В том, что эти файлы уже добавлены, можно убедиться и вручную, увидев нужную конфигурацию до того, как будут исполнены следующие строки с ручным добавлением:

![image](https://github.com/HoroshoNaVolge/dotnetLearning/assets/20906347/e8687c95-8271-4081-bc46-042ac337de90)

navferty commented on src/AspNetCore.WebApi/Program.cs:
```diff
@@ -0,0 +1,52 @@
+
+            builder.Services.AddHttpClient("DadataClient", client =>
+            {
+                var options = builder.Configuration.GetSection("Dadata").Get<DadataServiceOptions>();
```
Чтобы не прописывать всякий раз имя секции, можно определить ее как константу. Удобно делать это в том же классе опций константой `SectionName`:

```c#
class DadataServiceOptions
{
  public const string SectionName = "Dadata";

  public string? DaDataApiBaseUrl { get; set; }
  public string? DaDataApiToken { get; set; }
}
```

navferty commented on src/AspNetCore.WebApi/Services/DadataService.cs:
```diff
@@ -0,0 +1,42 @@
+            var httpClient = _httpClientFactory.CreateClient("DadataClient");
+            var response = await httpClient.GetAsync($"?query={inn}");
+
+            if (response.IsSuccessStatusCode)
```
Рекомендую обратить внимание на паттерн [return early](https://medium.com/swlh/return-early-pattern-3d18a41bba8).

Такой подход поможет уменьшить вложенность кода, и сделать его более удобным для чтения. Вместо:

```c#
var result = DoSmth();
if (IsOk(result)
{
  // continue processing the result
}
// process non-ok result
throw new BlaBlaBlaException("ololo");
```

получится так:

```c#
var result = DoSmth();
if (!IsOk(result)
{
  // process non-ok result
  throw new BlaBlaBlaException("ololo");
}
// continue main scenario of processing the result
```

navferty commented on src/AspNetCore.WebApi/Services/DadataService.cs:
```diff
@@ -0,0 +1,42 @@
+            if (response.IsSuccessStatusCode)
+            {
+                using var stream = await response.Content.ReadAsStreamAsync();
+                using var jsonDocument = await JsonDocument.ParseAsync(stream);
```
Если структура ответа известна заранее, обычно нет необходимости в промежуточном парсинге в `JsonDocument`. Создаётся соответствующая модель в виде C# классов, и происходит десериализация в эту модель.

Кстати, имея пример ответа в виде JSON, можно воспрользоваться фичой студии "Paste JSON as classes":

![image](https://github.com/HoroshoNaVolge/dotnetLearning/assets/20906347/8c2eb309-cf83-48a2-9bd0-fac54a37efec)

navferty commented on src/AspNetCore.WebApi/Services/DadataService.cs:
```diff
@@ -0,0 +1,42 @@
+                    var firstSuggestion = suggestionsArray.FirstOrDefault();
+                    return firstSuggestion.GetProperty("value").GetString();
+                }
+                catch { return null; }
```
Крайне не рекомендуется использовать пустой блок catch и при этом не указывать ожидаемый тип исключения. Если нет возможности избежать выбрасывания исключения (например, при вызове стороннего кода), обязательно указать конкретный тип того исключения, которое может возникнуть. Или, хотя бы, какие-то конкретные условия на это исключение (и всё же залогировать факт этого исключения), например:

```c#
Result? result = null;
try
{
  result = DoSomeDangerousThingThatRandomlyThrows();
}
catch (Exception ex) when (ex.Message.Contains("Бредовая ошибка в чужом коде, которая иногда возникает по непонятным причинам, и которую мы не можем сами исправить"))
{
  _logger.LogWarning("опять эта ошибка: {Error}", ex.Message);
}

// process result
```

navferty commented on src/AspNetCore.WebApi/Services/DadataService.cs:
```diff
@@ -0,0 +1,42 @@
+            _httpClientFactory = httpClientFactory;
+        }
+
+        public async Task<string?> GetOrganizationNameByInnAsync(string inn)
```
Этот сервис, теоретически, может использоваться не только для простого вывода текста пользователю, но и, например, быть переиспользован в более сложном коде, где, например, результат будет записан в БД. Для этого может быть намного удобнее, если об ошибке будет сообщено явно. Это может быть либо выбрасывание собственного исключения, либо возврат своего типа, вроде

```c#
class QueryOrganizationResult
{
  bool IsSuccess {get;set;}
  string? ErrorDescription {get;set;}
  string? OrganizationName {get;set;}
}
```

Кроме того, если вместо простой строки сервис возвращает такой класс, решение будет намного проще расширять. Скажем, если наряду с именем организации нужно будет возвращать её адрес, или количество полученных suggestion'ов - не придётся менять сигнатуру этого метода, а лишь добавить новые свойства в возвращаемый тип резалт. Тем самым не придётся ломать код, в котором уже есть вызовы этого метода.

navferty commented on src/AspNetCore.WebApi/Controllers/DadataController.cs:
```diff
@@ -0,0 +1,27 @@
+        }
+
+        [HttpGet(template: "get")]
+        public async Task<ActionResult<string>> GetOrganizationInfo(string inn)
```
Рекомендую в API-контроллерах добавлять аргумент `CancellationToken cancellationToken`. Этот токен отмены прокидывается во все асинхронные вызовы. В случае, если клиент отменяет запрос, в ходе которого на бэкенде происходит, например, тяжелый запрос к БД, для экономии ресурсов лучше отменить эту операцию.

https://www.youtube.com/watch?v=b5dyPJ3zyRg&ab_channel=NickChapsas

Надо заметить, что в более сложных сценариях при отмене запроса может потребоваться всё равно довести всю операцию до конца, либо вызвать отдельный код для обработки этой ситуации: https://stackoverflow.com/a/58534523/16412442

HoroshoNaVolge committed:
Delete appsettings.Development.json

HoroshoNaVolge committed:
Merge pull request #10 from HoroshoNaVolge/deleteAppSettings.Dev.Json

Delete appsettings.Development.json

HoroshoNaVolge committed:
hideTokenIn Development.json

HoroshoNaVolge committed:
Merge pull request #11 from HoroshoNaVolge/hideToken

hideTokenIn Development.json

HoroshoNaVolge committed:
IntermediateCommit

Пока не работает:
1. Использование appsettings.Development.json в среде Development. Разобираюсь с Configuration.

2. Десериализация работает. Попробую сделать более абстрактно. Переписать метод и модель для возврата других данных, кроме названия компании

navferty commented on src/AspNetCore.WebApi/Program.cs:
```diff
@@ -0,0 +1,51 @@
+            builder.Services.AddHttpClient("DadataClient", client =>
+            {
+                var options = builder.Configuration.GetSection(DadataServiceOptions.SectionName).Get<DadataServiceOptions>();
+                client.BaseAddress = new Uri(options.DaDataApiBaseUrl);
```
В отличие от получения `IOptions<TOptions>` через DI, этот метод может вернуть null. Об этом также отображается подсказка в студии:

![image](https://github.com/HoroshoNaVolge/dotnetLearning/assets/20906347/7d56277b-5b2a-4604-9280-1b44a4ddb7fc)

Рекомендую проверить на null, и поскольку продолжение работы данного сервиса без этой секции конфигурации не имеет смысла, вывести сообщение об ошибке и выбросить исключение, так чтобы сообщение явно говорило о причине ошибки. В противном случае, при ошибках конфигурации будет малопонятное `NullReferenceException`, и придётся тратить больше времени на поиск проблемы.

navferty commented on src/AspNetCore.WebApi/Services/DadataService.cs:
```diff
@@ -0,0 +1,42 @@
+
+            else
+            {
+                using var stream = await response.Content.ReadAsStreamAsync();
```
Опять же, рекомендую внимательно относиться ко всем предупреждениям, предложениям и подсказкам (зелёные подчеркивания, а также слабый пунктир), которые даёт Visual Studio: в данном случае - проброс токена отмены в асинхронные вызовы `GetAsync` и `ReadAsStreamAsync`, а также проверка `result` на null:

![image](https://github.com/HoroshoNaVolge/dotnetLearning/assets/20906347/3654be8b-8365-49c3-a472-a368bed3bc81)

Если эти подсказки не видны в студии, возможно стоит уточнить её конфигурацию в меню Tools->Options->Text Editor->C#

navferty commented on .gitignore:
```diff
@@ -396,3 +396,5 @@ FodyWeavers.xsd
 # JetBrains Rider
 *.sln.iml
+/src/AspNetCore.WebApi/appsettings.Development.json
+/src/AspNetCore.WebApi/appsettings.json
```
appsettings.json лучше оставить закоммиченным.

Суть в том, чтобы в appsettings.json была вся структура конфигурации, включая примеры значений, но чувствительные данные (пароли, токены и т.д.) заменены на заглушку. В этом случае любой, кто склонирует себе проект, сможет иметь возможность сконфигурировать его.

При этом appsettings.Development.json, который исключён из системы контроля версий, будет лежать в локальной папке с проектом, может содержать всю необходимую конфигурацию. При запуске с опцией `-c Debug` значения из этого файла будут переопределять те, которые прописаны в основном файле appsettings.json

navferty commented on src/AspNetCore.WebApi/Models/QueryOrganizationResult.cs:
```diff
@@ -0,0 +1,14 @@
+﻿namespace AspNetCore.WebApi.Models
+{
+    public class QueryOrganizationResult
```
Я бы предложил разделить модели, которые используются для десериализации ответа от стороннего сервиса, и модели, которые отдаёт сам сервис. Даже если сейчас они совпадают, в какой-то момент API стороннего сервиса может поменяться - но это не должно повлиять на API нашего сервиса.

navferty replied:
Этот класс всё еще используется для десериализации: `JsonSerializer.Deserialize<QueryOrganizationResult>`, при том что он является частью публичного API в сервисе `IDadataService`.

Более того, по факту все его свойства, кроме одного, отсутствуют в ответе от стороннего сервиса. Я бы всё же предложил их разделить, причем модель, используемая для десериализации, может быть объявлена как `internal class` (на случай, если `IDadataService` и его реализация будет вынесен в отдельный проект-библиотеку классов. А вот тот тип, который сервис отдает в публичном методе, должен и сам быть публичным.

HoroshoNaVolge committed:
comments resolved (probably not)

С IOptions<TOptions> сломал голову, наверное это не так должно быть реализовано

navferty commented on .gitignore:
```diff
@@ -396,3 +396,6 @@ FodyWeavers.xsd
 # JetBrains Rider
 *.sln.iml
+/src/AspNetCore.WebApi/appsettings.Development.json
+/src/AspNetCore.WebApi/appsettings.json
```
appsettings.json также можно не добавлять в гитигнор

navferty commented on src/AspNetCore.WebApi/Program.cs:
```diff
@@ -0,0 +1,55 @@
+               var options = sp.GetRequiredService<IOptions<DadataServiceOptions>>().Value;
+               {
+                   if (options.DaDataApiBaseUrl == null)
+                       throw new ArgumentNullException("��� DaDataApiBaseUrl � ����� ������������ ��� � ��������� �� ��������������� ���� ������������");
```
кажется, проблемы с кодировкой при коммите. проверь кодировку локального файла (рекомендуется UTF-8) и настройки гит клиента

navferty commented on src/AspNetCore.WebApi/Program.cs:
```diff
@@ -0,0 +1,55 @@
+            builder.Services.AddHttpClient("DadataClient")
+               .ConfigureHttpClient((sp, client) =>
+           {
+               var options = sp.GetRequiredService<IOptions<DadataServiceOptions>>().Value;
```
В текущем варианте, эта лямбда-функция вызывается только при вызове `_httpClientFactory.CreateClient`, при обработке запроса. Это значит, что сам сервис сможет стартовать нормально, без ошибок, даже в случае если необходимая секция отсутствует в конфигах.

Я бы предложил получить экземпляр `DadataServiceOptions` и проверить его сразу, в самом теле метода `Main` - так ошибка станет видна сразу при попытке запустить сервис.

Насчёт `IOptions<T>` - он как правило используется если эти опции нужны где-то в классе сервиса, тогда этот тип просто запрашивается в конструкторе, наряду с другими зависимостями:

```c#
internal sealed class SomeService
{
  private readonly IFirstDependency dependency1;
  private readonly IOptions<SomeConfiguration> options;

  public SomeService(IFirstDependency dependency1, IOptions<SomeConfiguration> options)
  {
    this.dependency1 = dependency1;
    this.options= options;
  }

  // ...
}
```

А в данном конкретном случае, я бы и сам наверно сделал так же, как было в предыдущей итерации, через `GetSection(...).Get<...>()` (но тут дело вкуса).

navferty commented on src/AspNetCore.WebApi/Program.cs:
```diff
@@ -0,0 +1,55 @@
+           {
+               var options = sp.GetRequiredService<IOptions<DadataServiceOptions>>().Value;
+               {
+                   if (options.DaDataApiBaseUrl == null)
```
Если это строка, можно немного расширить проверку, используя метод `string.IsNullOrEmpty(value)`

navferty commented on src/AspNetCore.WebApi/Services/DadataService.cs:
```diff
@@ -0,0 +1,39 @@
+                using var stream = await response.Content.ReadAsStreamAsync(token);
+                var result = JsonSerializer.Deserialize<QueryOrganizationResult>(stream);
+
+                if (result == null || result?.Suggestions?.Count() == 0)
```
Так как ниже всё равно используется метод `FirstOrDefault`, который пытается итерировть коллекцию до первого элемента, можно немного упростить код (ведь `Count` так же делает итерацию, причем итерирует до конца):

```c#
var item = result?.Suggestions?.FirstOrDefault();
if (item is null)
  // ...
```

navferty replied:
Кроме того, в общем случае я бы рекомендовал избегать выражения вида

```c#
if (someEnumerable.Count() == 0)
```

предпочитая вариант:

```c#
if (someEnumerable.Any())
```

Ведь метод `Count()` делает в точности что его просят - подсчитывает общее количество элементов. Но в данном случае оно не важно, на результат выражения влияет только факт наличия хотя бы одного элемента.

Иными словами, нет необходимости проходить `foreach`-ем до конца всей коллекции, чтобы понять, что в ней что-то есть. Достаточно лишь проверить, что выполнится хотя бы одна итерация, после чего выйти из этого цикла.

HoroshoNaVolge committed:
comments resolved

navferty commented on src/AspNetCore.WebApi/Services/DadataService.cs:
```diff
@@ -0,0 +1,35 @@
+
+                return new() { IsSuccess = false, ErrorDescription = "В запросе использован невалидный API токен", OrganizationName = null };
+
+            else
```
Можно убрать блок `else` - ведь в первой ветке есть `return`. Это уменьшит вложенность кода.

HoroshoNaVolge committed:
Update README.md

navferty commented on src/AspNetCore.WebApi/Program.cs:
```diff
@@ -0,0 +1,66 @@
+
+            builder.Services.Configure<DadataServiceOptions>(builder.Configuration.GetSection(DadataServiceOptions.SectionName));
+
+            // Не совсем уверен, что правильно передавать options потом в ConfigureHttpClient из этого контекста.
```
В целом, не вижу ничего плохого. К слову, такой захват переменный происходит через замыкания (closure). Для информации:

* [статья на metanit'е](https://metanit.com/sharp/tutorial/3.54.php)
* [хорошая статья от pvs studio](https://pvs-studio.ru/ru/blog/posts/csharp/0468/)
* [пример для самостоятельного разбора](https://github.com/navferty/dotnet-learning?tab=readme-ov-file#%D0%BF%D1%80%D0%B8%D0%BC%D0%B5%D1%80-%D0%B4%D0%BB%D1%8F-%D1%80%D0%B0%D0%B7%D0%B1%D0%BE%D1%80%D0%B0), со ссылкой на sharplab с кодом

navferty commented on src/AspNetCore.WebApi/AspNetCore.WebApi.http:
```diff
@@ -0,0 +1,6 @@
+@AspNetCore.WebApi_HostAddress = http://localhost:5202
+
+GET {{AspNetCore.WebApi_HostAddress}}/weatherforecast/
```
Этот [файл](https://learn.microsoft.com/en-us/aspnet/core/test/http-files?view=aspnetcore-8.0) рекомендую либо привести в соответствие с API, либо удалить вовсе.

HoroshoNaVolge committed:
comments resolved

navferty commented on src/AspNetCore.WebApi/Models/Suggestion.cs:
```diff
@@ -2,8 +2,10 @@
+    internal class Suggestion
     {
+        [JsonPropertyName("suggestions")]
+        public Suggestion[]? Suggestions { get; set; }
```
Ответ точно имеет такую рекурсивную структуру? если нет - лучше не полениться создать иерархию из двух различных классов.

HoroshoNaVolge committed:
Suggestion model divided
